# nf-synteny 

## Installation

### Prerequistites

- [Docker](https://docs.docker.com/engine/install/) or [Singularity](https://docs.sylabs.io/guides/3.11/admin-guide/installation.html).
- [Java](https://www.java.com/en/download/help/download_options.html) and [openJDK](https://openjdk.org/install/) >= 8 (**Please Note:** When installing Java versions are `1.VERSION` so `Java 8` is `Java 1.8`).
- [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html) >= `v22.03.0-edge`.

### Install

To install the pipeline please use the following commands but replace VERSION with a [release](https://github.com/Eco-Flow/synteny/releases).

`wget https://github.com/Eco-Flow/synteny/archive/refs/tags/VERSION.tar.gz -O - | tar -xvf -`

or

`curl -L https://github.com/Eco-Flow/synteny/archive/refs/tags/VERSION.tar.gz --output - | tar -xvf -`

This will produce a directory in the current directory called `synteny-VERSION` which contains the pipeline.

## Inputs

### Required

* `--input /path/to/csv/file` - A singular csv file as input in one of the two formats stated below.

This csv can take 2 forms:
* A 2 field csv where each row is a unique species name followed by a Refseq genome reference ID (**NOT** a Genbank reference ID) i.e. `data/Example-accession.csv`. The pipeline will download the relevant genome fasta file and annotation gff3 (or gff augustus) file.
* A 3 field csv where each row is a unique species name, followed by an absolute path to a genome fasta file, followed by an absolute path to an annotation gff3 (or gff augustus) file i.e. `data/Example-local.csv`.

**Please Note:** The genome has to be chromosome level not contig level.

### Optional 

* `--outdir /path/to/output/directory` - A path to the output directory where the results will be written to (**Default:** `Results`).
* `--hex /path/to/hex/file` - A path to a file containing a singular, unique hex code on each line to be used when painting chromosomes (**Default:** `data/unique_hex`).
* `--go /path/to/directory/containing/species/hash/files` - A path to a directory containing a hash file for each species in the analysis i.e. `data/go_input/hash_files`. These hash files can be generated by running the [Goatee](https://github.com/chriswyatt1/Goatee) pipeline.
* `--tree /path/to/tree/file` - A path to a file containing a phylogenetic tree for all species in Newick format i.e. `data/score_tree_input/tree.txt`.
* `--clean` - A true or false value assigned to this parameter will determine whether the work directory is automatically deleted or not if the pipeline is successful. Deleting the work directory saves space however it will not be possible to use this work directory in future for caching (**Default:** `true`).
* `--architecture` - An `amd` or `arm` value assigned to this parameter determines whether containers built for the amd or arm CPU architecture are used (**Default:** `amd`).

## Profiles

This pipeline is designed to run in various modes that can be supplied as a comma separated list i.e. `-profile profile1,profile2`.

### Container Profiles

Please select one of the following profiles when running the pipeline.

* `docker` - This profile uses the container software Docker when running the pipeline. This container software requires root permissions so is used when running on cloud infrastructure or your local machine (depending on permissions). **Please Note:** You must have Docker installed to use this profile.
* `singularity` - This profile uses the container software Singularity when running the pipeline. This container software does not require root permissions so is used when running on on-premise HPCs or you local machine (depending on permissions). **Please Note:** You must have Singularity installed to use this profile.

### Optional Profiles

* `local` - This profile is used if you are running the pipeline on your local machine.
* `test` - This profile is used if you want to test running the pipeline on your infrastructure. **Please Note:** You do not provide any input parameters if this profile is selected but you still provide a container profile.

## Custom Configuration

If you want to run this pipeline on your institute's on-premise HPC or specific cloud infrastructure then please contact us and we will help you build and test a custom config file. This config file will be published to our [configs repository](https://github.com/Eco-Flow/configs). 

## Running the Pipeline

**Please note:** The `-resume` flag uses previously cached successful runs of the pipeline.

* Running the pipeline with local and Docker profiles:
`nextflow run main.nf -profile docker,local -resume --input data/Example-accession.csv`

* Running the pipeline with Singularity and test profiles:
`nextflow run main.nf -profile singularity,test`

* Running the pipeline with all parameters:
`nextflow run main.nf -profile singularity,local -resume --input data/Example-local.csv --clean false --architecture arm --go data/go_input/hash_files --tree data/score_tree_input/tree.txt`

* Running the pipeline with a custom config file:
`nextflow run main.nf -profile docker,local -resume --input data/Example-accession.csv -c /path/to/custom/config`

## Results

Once completed, your output directory should have a folder called Jcvi_results, in which there should be a:

1. Dot plot (<Species1><Species2>.pdf). Showing the chromosome synteny as a dot plot.
2. Chromosome synteny plot (<Species1><Species2>.macro.pdf). Showing a 1 to 1 chromosome mapping with lines drawn between syntenic chromosomes.
3. Depth plot (<Species1><Species2>.depth.pdf). Number of 1to1 or 1toMany orthologs detected.
4. Chromosome painted mappings (). Showing on graphic chromosomes, which sections are syntenic between two species.
5. Anchors. (<Species1><Species2>.anchors). This shows the syntenic blocks, gene by gene.

All of the pipeline run information can be found inside `pipeline_info`.

The CO2 emissions of your pipeline can be found inside `co2_emissions`.

## Citation

Please cite : "Tang et al. (2008) Synteny and Collinearity in Plant Genomes. Science".

## Contact Us

If you need any support do not hesitate to contact us at any of:

`simon.murray [at] ucl.ac.uk`

`c.wyatt [at] ucl.ac.uk` 

`ecoflowucl [at] gmail.com`
